src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ShardStateCollector.java

```diff
@@ -0,0 +1,131 @@
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.collectors;
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.ESResources;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsConfiguration;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsProcessor;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.RoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.jooq.tools.StringUtils;
+import org.jooq.tools.json.JSONObject;
+
+import java.util.List;
+
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics.SHARD_PRIMARY;
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics.SHARD_REPLICA;
+
+public class ShardStateCollector extends PerformanceAnalyzerMetricsCollector implements MetricsProcessor {
+    public static final int SAMPLING_TIME_INTERVAL = MetricsConfiguration.CONFIG_MAP.get(ShardStateCollector.class).samplingInterval;
+    private static final Logger LOG = LogManager.getLogger(ShardStateCollector.class);
+    private static final int KEYS_PATH_LENGTH = 0;
+    private StringBuilder value;
+
+    public ShardStateCollector() {
+        super(SAMPLING_TIME_INTERVAL, "ShardsStateCollector");
+        value = new StringBuilder();
+    }
+
+    @Override
+    void collectMetrics( long startTime) {
```

Comment:
We also want to track how long we spend inside the `collectMetrics`

Reply:
The change is done for all the collectors in https://github.com/opendistro-for-elasticsearch/performance-analyzer-rca/pull/436/files#diff-115bbe9019d03c22dc720eb9fbc6670f354808d16bebb1eb6933dd468ce470af